### PR TITLE
#151 Single Config

### PR DESCRIPTION
**Pull request by hunterwilliams**
For issue #151

### CODE REVIEW

File: app/templates/README.mdown [hunk 1]
```diff
@@ -42,9 +42,9 @@ On Windows:
 
     ml.bat local deploy content
 
-Edit `gulp.config.js` to set your desired ports
+Edit `./deploy/gulp-local.json` to set your desired ports
```
Comment (grtjn) [inline app/templates/README.mdown]: you can put the gulp-local.json in the root, instead of under deploy/

File: app/templates/gulpfile.js [hunk 2]
```diff
@@ -473,17 +482,27 @@ function serve(isDev, specRunner) {
...
+ * Determine whether the enviornment requires dev mode'
+ * @param {String} env - 'local', 'dev', or 'prod'
+ * @return {Boolean} - true if env==='local'
+ */
+function isDevMode(env) {
+  return env === 'local';
+}
+
+function getNodeOptions(env) {
+  var envJson = require('./deploy/gulp-' + env + '.json');
```
Comment (grtjn) [inline app/templates/gulpfile.js]: Won't the require break if gulp-local.json doesn't exist? If yes, perhaps try to ignore the error, and maybe print a message suggesting you can provide local overrides with that file.

File: slushfile.js [hunk 3]
```diff
@@ -260,34 +273,14 @@ gulp.task('generateSecret', ['init'], function(done) {
 gulp.task('configGulp', ['init'], function(done) {
 
   try {
+    var configJSON = {};
+    configJSON['ml-host'] = 'localhost';
```
Comment (grtjn) [inline slushfile.js]: Can you make the slush generator ask for the host as well?

**Comment by hunterwilliams** [comment]
@grtjn I've added the change so it doesn't need the xccport for ml8.

also I renamed ml-port to ml-http-port because it's a good clarification I think for inexperienced users

File: app/templates/gulpfile.js [hunk 4]
```diff
@@ -439,12 +448,12 @@ function orderSrc (src, order) {
  * serve the code
  * --debug-brk or --debug
  * --nosync
- * @param  {Boolean} isDev - dev or build mode
+ * @param  {String} env - 'dev','test', or 'prod'
```
Comment (grtjn) [inline app/templates/gulpfile.js]: Small typo? env = local | dev | prod..

File: slushfile.js [hunk 5]
```diff
@@ -260,34 +289,19 @@ gulp.task('generateSecret', ['init'], function(done) {
...
+    if (settings.mlVersion < 8) {
+      configJSON['ml-xcc-port'] = settings.xccPort;
+    }
 
-    fs.writeFileSync('etc/init.d/node-express-service', nodeService);
+    var configString = JSON.stringify(configJSON, null, 2) + '\n';
+    fs.writeFileSync('gulp-local.json', configString, { encoding: 'utf8' });
   } catch (e) {
-    console.log('failed to update node-express-service: ' + e.message);
+    console.log('failed to write gulp.local.json: ' + e.message);
```
Comment (grtjn) [inline slushfile.js]: Tiny typo, gulp-local.json..

File: app/templates/gulpfile.js [hunk 6]
```diff
@@ -473,17 +482,35 @@ function serve(isDev, specRunner) {
...
+
+function getNodeOptions(env) {
+  var envJson;
+  var envFile = 'gulp-' + env + '.json';
+    try {
+      envJson = require(envFile);
+    }
+    catch (e) {
+      envJson = {};
+      console.log('Info: ' + envFile + ' - may not exist. Did you mean to create one?');
```
Comment (grtjn) [inline app/templates/gulpfile.js]: A different project was using a similar pattern for a `local.gradle` file. If missing it printed:

"Couldn't find local.gradle; you can create this gitignore'd file to override properties if necessary"

Maybe print a similar message in our case?

Comment (grtjn) [inline app/templates/gulpfile.js]: A thought just struck me. People will ask themselves, how should gulp-local.json look like? How about adding a gulp init-local task that helps create that file, and refer to that in the info message?

File: slushfile.js [hunk 7]
```diff
@@ -260,34 +289,19 @@ gulp.task('generateSecret', ['init'], function(done) {
 gulp.task('configGulp', ['init'], function(done) {
 
   try {
+    var configJSON = {};
+    configJSON['ml-host'] = settings.marklogicHost;
```
Comment (grtjn) [inline slushfile.js]: tiny favor: can you add ml-version in here too?

**Comment by grtjn** [comment]
Good work, really looking forward to this.. :)

**Comment by hunterwilliams** [comment]
This does not cover proxy.js and routes.js properly for port changing looking into that now

**Comment by hunterwilliams** [comment]
Fixed that. The change to of env.ML_PORT to env.ML_HTTP_PORT broke it as I wasn't populating it in getNodeOptions(). I changed it back since in getNodeOptions() it looks for env.ML_PORT and I'm not sure where that value is supposed to come from

Thoughts?

**Comment by grtjn** [comment]
env.ML_PORT is used between gulpfile.js and node-app.js mostly. Looks good enough as is.

There seems to be a merge conflict, can you rebase? I'll have a spin with these changes after rebase..

**Comment by grtjn** [comment]
Yay, didn't need to change generated local.properties and gulp-local.json to run ml local bootstrap and gulp serve! Really happy with this..

Just need to remind myself to not run gulp serve-dev, but `gulp serve-local` from now on!

**Pull request merged**